src/oidcc_token.erl

```diff
@@ -349,6 +356,130 @@ retrieve(AuthCode, ClientContext, Opts) ->
+        response_modes_supported = ResponseModesSupported,
+        authorization_signing_alg_values_supported = SigningAlgSupported0,
+        authorization_encryption_alg_values_supported = EncryptionAlgSupported0,
+        authorization_encryption_enc_values_supported = EncryptionEncSupported0
+    } =
+        Configuration,
+
+    case
+        lists:member(<<"jwt">>, ResponseModesSupported) or
+            lists:member(<<"query.jwt">>, ResponseModesSupported) or
+            lists:member(<<"form_post.jwt">>, ResponseModesSupported)
+    of
+        true ->
+            SigningAlgSupported =
+                case SigningAlgSupported0 of
+                    undefined -> [];
+                    SigningAlgs -> SigningAlgs
+                end,
+            EncryptionAlgSupported =
+                case EncryptionAlgSupported0 of
+                    undefined -> [];
+                    EncryptionAlgs -> EncryptionAlgs
+                end,
+            EncryptionEncSupported =
+                case EncryptionEncSupported0 of
+                    undefined -> [];
+                    EncryptionEncs -> EncryptionEncs
+                end,
+            JwksWithClientJwks =
+                case ClientJwks of
+                    none -> Jwks;
+                    #jose_jwk{} -> oidcc_jwt_util:merge_jwks(Jwks, ClientJwks)
+                end,
+
+            SigningJwks =
+                case oidcc_jwt_util:client_secret_oct_keys(SigningAlgSupported, ClientSecret) of
+                    none ->
+                        JwksWithClientJwks;
+                    SigningOctJwk ->
+                        oidcc_jwt_util:merge_jwks(JwksWithClientJwks, SigningOctJwk)
+                end,
+            EncryptionJwks =
+                case oidcc_jwt_util:client_secret_oct_keys(EncryptionAlgSupported, ClientSecret) of
+                    none ->
+                        JwksWithClientJwks;
+                    EncryptionOctJwk ->
+                        oidcc_jwt_util:merge_jwks(JwksWithClientJwks, EncryptionOctJwk)
+                end,
+            %% https://openid.net/specs/oauth-v2-jarm-final.html#name-processing-rules
+            %% 1. decrypt if necessary
+            %% 2. validate <<"iss">> claim
+            %% 3. validate <<"aud">> claim
+            %% 4. validate <<"exp">> claim
+            %% 5. validate signature (valid, not <<"none">> alg)
+            %% 6. continue processing
+            maybe
+                {ok, DecryptedResponse} ?=
+                    oidcc_jwt_util:decrypt_if_needed(
+                        Response,
+                        EncryptionJwks,
+                        EncryptionAlgSupported,
+                        EncryptionEncSupported
+                    ),
+                ExpClaims = [
+                    {<<"iss">>, Issuer}
+                ],
+                TrustedAudience = maps:get(trusted_audiences, Opts, any),
+                {ok, #jose_jwt{fields = PeekClaims}} ?=
```

Comment:
Why are you validating peeked claims instead of `DecryptedResponse`?

Reply:
`DecryptedResponse` is still a binary. We validate against `PeekClaims` (rather than getting the claims after calling `oidcc_jwt_util:verify_signature/3`) because the spec says that the client must validate some of the claims before validating the signature: https://openid.net/specs/oauth-v2-jarm-final.html#name-processing-rules

Reply:
@paulswartz I'm not sure I understand your reasoning.

Currently the code is:

* decrypting
* peeking & validating peeked claims
* verifying & unpacking the decrypted token
* returning the verified claims

Why are we not:

* decypting
* verifying token & extracting claims
* check the verified claims

Do you read the spec as having a defined order to the steps? I doesn't specifically say to do so as far as I can see. Might be wrong on this)

I also don't see any hints that peeking is expected nor do I see a good reason to do so.

Reply:
Here's the text from https://openid.net/specs/oauth-v2-jarm-final.html#section-2.4-2
> The client MUST process the JWT secured response as follows:
> 1. (OPTIONAL) The client decrypts the JWT using the default key for the respective issuer or, if applicable, determined by the kid JWT header parameter. The key might be a private key, where the corresponding public key is registered with the expected issuer of the response ("use":"enc" via the client's metadata jwks or jwks_uri) or a key derived from its client secret (see [Section 2.2](https://openid.net/specs/oauth-v2-jarm-final.html#signing-and-encryption)).
> 2. The client obtains the iss element from the JWT and checks whether its value is well known and identifies the expected issuer of the authorization process in examination. If the check fails, the client MUST abort processing and refuse the response.
> 3. The client obtains the aud element from the JWT and checks whether it matches the client id the client used to identify itself in the corresponding authorization request. If the check fails, the client MUST abort processing and refuse the response.
> 4. The client checks the JWT's exp element to determine if the JWT is still valid. If the check fails, the client MUST abort processing and refuse the response.
> 5. The client MUST check the signature of the JWT according to [[RFC7515](https://openid.net/specs/oauth-v2-jarm-final.html#RFC7515)] and the algorithm none ("alg":"none") MUST NOT be accepted. If the check fails, the client MUST abort processing and refuse the response.

My read of this is:
1. decrypt if necessary
2. check the "iss" claim
3. check the "aud" claim
4. check the "exp" claim
5. check the signature and verify that "alg": "none" was not used

Since steps 2, 3, and 4 require checking the claims before the signature has been checked in step 5, we need to peek into the JWT to do that.

Reply:
Tracked via https://github.com/erlef/oidcc/issues/325